src/app/ReadHandler.cpp

```diff
@@ -0,0 +1,238 @@
+    mCurrentPriority  = reporting::PriorityLevel::Invalid;
+    mSuppressResponse = true;
+    mGetToAllEvents   = true;
+    mState            = HandlerState::Initialized;
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadHandler::Shutdown()
+{
+    ClearExistingExchangeContext();
+    MoveToState(HandlerState::Uninitialized);
+    mpExchangeMgr = nullptr;
+    mpDelegate    = nullptr;
+}
+
+CHIP_ERROR ReadHandler::ClearExistingExchangeContext()
+{
+    if (mpExchangeCtx != nullptr)
+    {
+        mpExchangeCtx->Abort();
+        mpExchangeCtx = nullptr;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void ReadHandler::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                                    const PayloadHeader & aPayloadHeader, System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferHandle response;
+
+    mpExchangeCtx = apExchangeContext;
+
+    err = ProcessReadRequest(std::move(aPayload));
+    SuccessOrExit(err);
+
+exit:
+    ChipLogFunctError(err);
+    return;
+}
+
+CHIP_ERROR ReadHandler::SendReportData(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    err = mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::ReportData, std::move(aPayload),
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kNone));
+    SuccessOrExit(err);
+    MoveToState(HandlerState::ReportDataSending);
+
+exit:
+    Shutdown();
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR ReadHandler::ProcessReadRequest(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferTLVReader reader;
+
+    ReadRequest::Parser readRequestParser;
+    EventPathList::Parser eventPathListParser;
+    chip::TLV::TLVReader eventPathListReader;
+
+    reader.Init(std::move(aPayload));
+
+    err = reader.Next();
+    SuccessOrExit(err);
+
+    err = readRequestParser.Init(reader);
+    SuccessOrExit(err);
+
+    err = readRequestParser.CheckSchemaValidity();
+    SuccessOrExit(err);
+
+    err = readRequestParser.GetEventPathList(&eventPathListParser);
+    SuccessOrExit(err);
+
+    eventPathListParser.GetReader(&eventPathListReader);
+
+    while (CHIP_NO_ERROR == (err = eventPathListReader.Next()))
+    {
+        VerifyOrExit(chip::TLV::AnonymousTag == eventPathListReader.GetTag(), err = CHIP_ERROR_INVALID_TLV_TAG);
+
+        EventPath::Parser eventPath;
+
+        err = eventPath.Init(eventPathListReader);
+        SuccessOrExit(err);
+
+        // TODO: Process and offer EventPath for RE to filter interested path
+    }
+
+    // if we have exhausted this container
+    if (CHIP_END_OF_TLV == err)
+    {
+        err = CHIP_NO_ERROR;
+    }
+
+    InteractionModelEngine::GetInstance()->GetReportingEngine()->ScheduleRun();
```

Comment:
@bzbarsky-apple , here it is scheduleRun to schedule report engine to send out report data to iniatiator